crates/compiler/gen_wasm/src/low_level.rs

```diff
@@ -2798,20 +2958,13 @@ fn list_map_n<'a>(
     }
     cb.i32_const(elem_ret_size as i32);
 
-    // If we have lists of different lengths, we may need to decrement
-    if arg_elem_layouts.len() > 1 {
-        for el in arg_elem_layouts.iter() {
-            // The dec function will be passed a pointer to the element within the list, not the element itself!
-            // Here we wrap the layout in a Struct to ensure we get the right code gen
-            let el_ptr = backend
-                .layout_interner
-                .insert_direct_no_semantic(LayoutRepr::Struct(backend.env.arena.alloc([*el])));
-            let idx = backend.get_refcount_fn_index(el_ptr, HelperOp::Dec);
-            let ptr = backend.get_fn_ptr(idx);
-            backend.code_builder.i32_const(ptr);
-        }
-    };
+    // We need to be able to increment the refcount of elements loaded.
+    for el in arg_elem_layouts.iter() {
+        let ptr = build_refcount_element_fn(backend, *el, HelperOp::IndirectInc);
+        backend.code_builder.i32_const(ptr);
+    }
```

Comment:
I see a comment in the old code about decrementing. I think for a `map2` on two lists of different lengths, the last few elements of the long list end up with a different refcount than the elements that have a corresponding entry in the short list.

Do we not need to do that any more? I have not been keeping up with the discussions 100% so maybe something has changed.

Reply:
Yeah, with the new setup, we don't decrement the tail anymore.

In the new setup: The original List holds 1 reference to each element. For each element we copy out of the original list, we must increment the refcount of when copying the element. When the original list is deallocated, it will cleanup the tail.